File: Offline/OfflinePlayer/objc/OfflinePlayer/ManagersUtilities/DownloadManager.m
//
//  DownloadManager.m
//  OfflinePlayer
//
//  Created by Jeremy Blaker on 10/30/19.
//  Copyright © 2019 Brightcove. All rights reserved.
//

#import "DownloadManager.h"

#import "BCOVVIdeo+Convenience.h"
#import "SettingsAdapter.h"

@interface DownloadManager ()

@property (nonatomic, assign) BOOL downloadInProgress;
@property (nonatomic, strong, readwrite) NSArray<BCOVOfflineVideoToken> *offlineVideoTokenArray;

@end

@implementation DownloadManager

+ (instancetype)sharedInstance
{
    static dispatch_once_t onceToken;
    static DownloadManager *downloadManager;
    dispatch_once(&onceToken, ^{
        downloadManager = [DownloadManager new];
    });
    return downloadManager;
}

#pragma mark - Private Methods

- (instancetype)init
{
    if (self = [super init])
    {
        [self setup];
    }
    return self;
}

- (void)setup
{
    NSDictionary *optionsDictionary =
    @{
      kBCOVOfflineVideoManagerAllowsCellularDownloadKey: @(NO),
      kBCOVOfflineVideoManagerAllowsCellularPlaybackKey: @(NO),
      kBCOVOfflineVideoManagerAllowsCellularAnalyticsKey: @(NO)
      };
    [BCOVOfflineVideoManager initializeOfflineVideoManagerWithDelegate:self
                                                               options:optionsDictionary];
    self.offlineVideoManager = BCOVOfflineVideoManager.sharedManager;
    
    self.videoPreloadQueue = @[].mutableCopy;
    self.videoDownloadQueue = @[].mutableCopy;
}

- (void)notifyDelegateToRefreshUI
{
    if ([self.delegate respondsToSelector:@selector(shouldRefreshUI)])
    {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.delegate shouldRefreshUI];
        });
    }
}

// Alert and return YES if the video is already downloaded or in a queue
- (BOOL)videoAlreadyProcessing:(BCOVVideo *)video
{
    BCOVOfflineVideoManager *ovm = BCOVOfflineVideoManager.sharedManager;

    // First check to see if the video is in a preload queue
    // videoPreloadQueue is an array of NSDictionary objects,
    // with a BCOVVideo under each "video" key.
    for (NSDictionary *videoDictionary in DownloadManager.sharedInstance.videoPreloadQueue)
    {
        BCOVVideo *testVideo = videoDictionary[@"video"];
        
        if ([video videoMatchesVideo:testVideo])
        {
            if ([self.delegate respondsToSelector:@selector(videoAlreadyPreloadQueued:)])
            {
                [self.delegate videoAlreadyPreloadQueued:video];
            }
            return YES;
        }
    }
    
    // First check to see if the video is in a download queue
    // videoDownloadQueue is an array of BCOVVideo objects
    for (NSDictionary *videoDownloadDictionary in DownloadManager.sharedInstance.videoDownloadQueue)
    {
        BCOVVideo *testVideo = videoDownloadDictionary[@"video"];
        if (testVideo == nil)
        {
            continue;
        }

        if ([video videoMatchesVideo:testVideo])
        {
            if ([self.delegate respondsToSelector:@selector(videoAlreadyDownloadQueued:)])
            {
                [self.delegate videoAlreadyDownloadQueued:video];
            }
            return YES;
        }
    }
    
    // Next check to see if the video has already been downloaded
    // or is in the process of downloading
    NSArray<BCOVOfflineVideoToken> *offlineVideoTokens = ovm.offlineVideoTokens;
    
    for (BCOVOfflineVideoToken offlineVideoToken in offlineVideoTokens)
    {
        BCOVVideo *testVideo = [ovm videoObjectFromOfflineVideoToken:offlineVideoToken];
        
        if ([video videoMatchesVideo:testVideo])
        {
            BCOVOfflineVideoStatus *status = [ovm offlineVideoStatusForToken:offlineVideoToken];
            
            // If the status is error, alert the user and allow them to retry the download
            if (status.downloadState == BCOVOfflineVideoDownloadStateError)
            {
                if ([self.delegate respondsToSelector:@selector(videoPreviouslyFailedDownload:offlineVideoToken:)])
                {
                    [self.delegate videoPreviouslyFailedDownload:video offlineVideoToken:offlineVideoToken];
                }
            }
            else
            {
                if ([self.delegate respondsToSelector:@selector(videoAlreadyDownloaded:)])
                {
                    [self.delegate videoAlreadyDownloaded:video];
                }
            }
            
            return YES;
            
        }
    }
    
    return NO;
}

- (void)runPreloadVideoQueue
{
    NSDictionary *videoDownloadDictionary = self.videoPreloadQueue.firstObject;
    BCOVVideo *video = videoDownloadDictionary[@"video"];
    NSDictionary *parameters = videoDownloadDictionary[@"parameters"];
    
    // Once the preload queue is empty, start the download queue
    if (video == nil)
    {
        [self downloadVideoFromQueue];
        return;
    }
    
    [self.videoPreloadQueue removeObject:videoDownloadDictionary];
    
    // Preloading only applies to FairPlay-protected videos.
    // If there's no FairPlay involved, the video is moved on
    // to the video download queue.
    if (!video.usesFairPlay)
    {
        NSLog(@"Video \"%@\" does not use FairPlay; preloading not necessary", video.properties[kBCOVVideoPropertyKeyName]);
        [self.videoDownloadQueue addObject:videoDownloadDictionary];
        
        NSAssert(NSThread.isMainThread, @"Must update UI on main thread");
        [self notifyDelegateToRefreshUI];

        [self runPreloadVideoQueue];
    }
    else
    {
        __weak typeof(self) weakSelf = self;
        [BCOVOfflineVideoManager.sharedManager preloadFairPlayLicense:video
                                                           parameters:parameters
         
                                                           completion:^(BCOVOfflineVideoToken offlineVideoToken, NSError *error) {
                                                               
                                                               dispatch_async(dispatch_get_main_queue(), ^{
                                                                   
                                                                   __strong typeof(weakSelf) strongSelf = weakSelf;
                                                                   
                                                                   if (error == nil)
                                                                   {
                                                                       NSLog(@"Preloaded %@", offlineVideoToken);
                                                                       
                                                                       [strongSelf.videoDownloadQueue addObject:videoDownloadDictionary];
                                                                       
                                                                       [strongSelf notifyDelegateToRefreshUI];
                                                                   }
                                                                   else
                                                                   {
                                                                       if ([strongSelf.delegate respondsToSelector:@selector(encounteredErrorPreloading:forVideo:)])
                                                                       {
                                                                           BCOVVideo *_video = offlineVideoToken ? [BCOVOfflineVideoManager.sharedManager videoObjectFromOfflineVideoToken:offlineVideoToken] : video;
                                                                           [strongSelf.delegate encounteredErrorPreloading:error forVideo:_video];
                                                                       }
                                                                   }
                                                                   
                                                                   [strongSelf runPreloadVideoQueue];
                                                                   
                                                               });
                                                               
                                                           }];
    }
}

- (void)downloadVideoFromQueue
{
    // If we're already downloading, this will be called automatically
    // when the download is done
    // Only needed for pre-iOS 11.4 only which can only handle
    // One download at a time
    if (@available(iOS 11.4, *))
    {}
    else if (self.downloadInProgress)
    {
        return;
    }
    
    NSDictionary *videoDownloadDictionary = self.videoDownloadQueue.firstObject;
    BCOVVideo *video = videoDownloadDictionary[@"video"];
    NSDictionary *parameters = videoDownloadDictionary[@"parameters"];

    if (video == nil)
    {
        // done!
        return;
    }
    
    [self.videoDownloadQueue removeObject:videoDownloadDictionary];
    
    self.downloadInProgress = YES;

    // Display all available bitrates
    [BCOVOfflineVideoManager.sharedManager variantBitratesForVideo:(BCOVVideo *)video
                                                        completion:^(NSArray<NSNumber *> *bitrates, NSError *error)
     {

         NSLog(@"Variant Bitrates for video: %@", video.properties[kBCOVVideoPropertyKeyName]);
         for (NSNumber *bitrateNumber in bitrates)
         {
             // Make sure the array contains the correct objects
             if (! [bitrateNumber isKindOfClass:NSNumber.class])
             {
                 NSLog(@"bitrateNumber contains the wrong class: %@", bitrateNumber.class.description);
             }
             
             NSLog(@"\t%d", bitrateNumber.intValue);
         }
     
     }];

    __weak typeof(self) weakSelf = self;
    
    AVURLAsset *avURLAsset = [self.offlineVideoManager urlAssetForVideo:video error:nil];
    // If mediaSelections is `nil` the SDK will default to the AVURLAsset's `preferredMediaSelection`
    NSArray<AVMediaSelection *> *mediaSelections = mediaSelections = avURLAsset.allMediaSelections;
    
    AVMediaSelectionGroup *legibleMediaSelectionGroup = [avURLAsset mediaSelectionGroupForMediaCharacteristic:AVMediaCharacteristicLegible];
    AVMediaSelectionGroup *audibleMediaSelectionGroup = [avURLAsset mediaSelectionGroupForMediaCharacteristic:AVMediaCharacteristicAudible];
    
    int counter = 0;
    for (AVMediaSelection *s in mediaSelections)
    {
        AVMediaSelectionOption *legibleMediaSelectionOption = [s selectedMediaOptionInMediaSelectionGroup:legibleMediaSelectionGroup];
        AVMediaSelectionOption *audibleMediaSelectionOption = [s selectedMediaOptionInMediaSelectionGroup:audibleMediaSelectionGroup];

        NSLog(@"AVMediaSelection option %i | legible display name: %@", counter, legibleMediaSelectionOption.displayName ?: @"nil");
        NSLog(@"AVMediaSelection option %i | audible display name: %@", counter, audibleMediaSelectionOption.displayName ?: @"nil");
        
        counter++;
    }

    [self.offlineVideoManager requestVideoDownload:video
                                   mediaSelections:mediaSelections
                                        parameters:parameters
                                        completion:^(BCOVOfflineVideoToken offlineVideoToken, NSError *error) {
                                            
                                            __strong typeof(weakSelf) strongSelf = weakSelf;
                                            dispatch_async(dispatch_get_main_queue(), ^{
                                                
                                                if (error == nil)
                                                {
                                                    // Success! Update our table with the new download status
                                                    if ([strongSelf.delegate respondsToSelector:@selector(videoDidBeginDownloading)])
                                                    {
                                                        [strongSelf.delegate videoDidBeginDownloading];
                                                    }
                                                }
                                                else
                                                {
                                                    strongSelf.downloadInProgress = NO;

                                                    // try again with another video
                                                    if (@available(iOS 11.4, *))
                                                    {}
                                                    else
                                                    {
                                                        [strongSelf downloadVideoFromQueue];
                                                    }

                                                    if ([strongSelf.delegate respondsToSelector:@selector(encounteredErrorDownloading:forVideo:)])
                                                    {
                                                        // Report any errors
                                                        BCOVVideo *video = [BCOVOfflineVideoManager.sharedManager videoObjectFromOfflineVideoToken:offlineVideoToken];
                                                        [strongSelf.delegate encounteredErrorDownloading:error forVideo:video];
                                                    }
                                                   
                                                }
                                                
                                            });
                                            
                                        }];
}

- (NSArray *)languagesArrayForAlternativeRenditions:(NSArray<NSDictionary *> *)alternativeRenditionAttributesDictionariesArray
{
    // We want to download all subtitle/audio tracks
    
    if (alternativeRenditionAttributesDictionariesArray == nil)
    {
        return nil;
    }
    
    NSLog(@"Alternative Rendition Attributes Dictionaries:\n%@", alternativeRenditionAttributesDictionariesArray);
    
    // Collect all the available subtitle languages in a set to avoid duplicates
    NSMutableSet *languagesSet = NSMutableSet.set;
    for (NSDictionary *alternativeRenditionAttributesDictionary in alternativeRenditionAttributesDictionariesArray)
    {
        NSString *typeString = alternativeRenditionAttributesDictionary[@"TYPE"];
        NSString *languageString = alternativeRenditionAttributesDictionary[@"LANGUAGE"];
        if ([typeString isEqualToString:@"SUBTITLES"] && languageString != nil)
        {
            [languagesSet addObject:languageString];
        }
    }
    
    NSArray *languagesArray = languagesSet.allObjects;
    
    {
        // For debugging: display the languages we found
        NSMutableString *languagesString = NSMutableString.string;
        BOOL first = YES;
        for (NSString *languageString in languagesArray)
        {
            // Add comma before each entry after the first
            if (first)
            {
                first = NO;
            }
            else
            {
                [languagesString appendString:@", "];
            }
            
            [languagesString appendString:languageString];
        }
        
        NSLog(@"Languages to download: %@", languagesString);
    }
    
    return languagesArray;
}

- (NSString *)mediaSelectionDescription:(AVMediaSelection *)mediaSelection
                      offlineVideoToken:(BCOVOfflineVideoToken)offlineVideoToken
{
    // Get the offline video object
    BCOVVideo *offlineVideo = [BCOVOfflineVideoManager.sharedManager videoObjectFromOfflineVideoToken:offlineVideoToken];
    
    // Get the path to the locally stored video and make an AVURLAsset out of it
    NSString *offlineVideoPath = offlineVideo.properties[kBCOVOfflineVideoFilePathPropertyKey];
    if (offlineVideoPath == nil)
    {
        return @"MediaSelection(n/a)";
    }

    NSURL *offlineVideoPathURL = [NSURL fileURLWithPath:offlineVideoPath];
    
    AVURLAsset *URLAsset = [AVURLAsset assetWithURL:offlineVideoPathURL];
    
    NSString *description = [self mediaSelectionDescription:mediaSelection
                                                   URLAsset:URLAsset];
    return description;
}

- (NSString *)mediaSelectionDescription:(AVMediaSelection *)mediaSelection
                               URLAsset:(AVURLAsset *)URLAsset
{
    // Return a string description of the specified Media Selection.
    AVMediaSelectionGroup *legibleMediaSelectionGroup = [URLAsset mediaSelectionGroupForMediaCharacteristic:AVMediaCharacteristicLegible];
    AVMediaSelectionGroup *audibleMediaSelectionGroup = [URLAsset mediaSelectionGroupForMediaCharacteristic:AVMediaCharacteristicAudible];
    AVMediaSelectionOption *legibleMediaSelectionOption = [mediaSelection selectedMediaOptionInMediaSelectionGroup:legibleMediaSelectionGroup];
    AVMediaSelectionOption *audibleMediaSelectionOption = [mediaSelection selectedMediaOptionInMediaSelectionGroup:audibleMediaSelectionGroup];
    
    NSString *description = [NSString stringWithFormat:@"MediaSelection(obj:%p, legible:%@, audible:%@)",
                             mediaSelection,
                             legibleMediaSelectionOption.displayName ?: @"-",
                             audibleMediaSelectionOption.displayName ?: @"-"];
    return description;
}


#pragma mark - Public Methods

- (void)removeOfflineToken:(BCOVOfflineVideoToken)offlineVideoToken
{
    NSMutableArray *updatedOfflineVideoTokenArray = self.offlineVideoTokenArray.mutableCopy;
    [updatedOfflineVideoTokenArray removeObject:offlineVideoToken];
    self.offlineVideoTokenArray = updatedOfflineVideoTokenArray;
}

- (NSMutableDictionary *)generateLicenseParameters
{
    NSMutableDictionary *licenseParameters = @{}.mutableCopy;

    // Generate the license parameters based on the Settings tab
    BOOL isPurchaseLicense = SettingsAdapter.purchaseLicenseType;
    // License details are only needed for FairPlay-protected videos.
    // It's harmless to add it for non-FairPlay videos too.
    
    if (isPurchaseLicense)
    {
        NSLog(@"Requesting Purchase License");
        licenseParameters[kBCOVFairPlayLicensePurchaseKey] = @YES;
    }
    else
    {
        unsigned long long rentalDuration = SettingsAdapter.rentalDuration;
        unsigned long long playDuration = SettingsAdapter.playDuration;
        
        NSLog(@"Requesting Rental License:\n"
              @"rentalDuration: %llu\n"
              @"playDuration: %llu\n",
              rentalDuration,
              playDuration);
        
        licenseParameters[kBCOVFairPlayLicenseRentalDurationKey] = @(rentalDuration);
        licenseParameters[kBCOVFairPlayLicensePlayDurationKey] = @(playDuration);
    }

    return licenseParameters;
}

- (NSMutableDictionary *)generateDownloadParameters
{    
    // Get base license parameters
    NSMutableDictionary *downloadParameters = [self generateLicenseParameters];

    // Add bitrate parameter for the primary download
    long long int bitrate = SettingsAdapter.bitrate;
    
    NSLog(@"Requested bitrate: %lld", bitrate);
    
    downloadParameters[kBCOVOfflineVideoManagerRequestedBitrateKey] = @(bitrate);

    return downloadParameters;
}

- (void)downloadVideo:(BCOVVideo *)video
{
    // See if the video has already been downloaded, or is pending download.
    // This displays an alert if necessary.
    if ([self videoAlreadyProcessing:video])
    {
        return;
    }

    NSMutableDictionary *downloadParameters = [self generateDownloadParameters];

    NSDictionary *videoDownloadDictionary = @{
        @"video": video,
        @"parameters": downloadParameters
    };
    
    [self.videoPreloadQueue addObject:videoDownloadDictionary];
    
    [self runPreloadVideoQueue];
}

- (void)retrieveVideoWithAccount:(NSString *)accountID
                         videoID:(NSString *)videoID
                      completion:(void (^)(BCOVVideo *video, NSDictionary *jsonResponse, NSError *error))completionHandler
{
    NSAssert(completionHandler!=nil, @"Completion handler cannot be nil");

    // Retrieve a playlist through the BCOVPlaybackService
    BCOVPlaybackServiceRequestFactory *playbackServiceRequestFactory = [[BCOVPlaybackServiceRequestFactory alloc] initWithAccountId:kDynamicDeliveryAccountID
                                                                                                                          policyKey:kDynamicDeliveryPolicyKey];
    BCOVPlaybackService *playbackService = [[BCOVPlaybackService alloc] initWithRequestFactory:playbackServiceRequestFactory];

    [playbackService findVideoWithVideoID:videoID
                               parameters:nil
                               completion:^(BCOVVideo *video, NSDictionary *jsonResponse, NSError *error) {

        // Pass on to caller
        completionHandler(video, jsonResponse, error);

     }];
}

- (void)updateOfflineTokens
{
    // Refresh the list of downloaded videos from the offline video manager
    self.offlineVideoTokenArray = [self.offlineVideoManager offlineVideoTokens];
}

#pragma mark - BCOVOfflineVideoManagerDelegate Methods

- (void)didCreateSharedBackgroundSesssionConfiguration:(NSURLSessionConfiguration *)backgroundSessionConfiguration
{
    // Helps prevent downloads from appearing to sometimes stall
    backgroundSessionConfiguration.discretionary = NO;
}

- (void)offlineVideoToken:(BCOVOfflineVideoToken)offlineVideoToken
             downloadTask:(AVAssetDownloadTask *)downloadtask
            didProgressTo:(NSTimeInterval)percent
{
    // This delegate method reports progress for the primary video download
    NSLog(@"Offline download didProgressTo: %0.2f%% for token: %@", (float)percent, offlineVideoToken);
    
    dispatch_async(dispatch_get_main_queue(), ^{

        if ([self.delegate respondsToSelector:@selector(downloadDidProgressTo:)])
        {
            [self.delegate downloadDidProgressTo:percent];
        }

    });
}

- (void)offlineVideoToken:(BCOVOfflineVideoToken)offlineVideoToken
didFinishDownloadWithError:(NSError *)error
{
    // The video has completed downloading
    
    NSLog(@"Download finished with error: %@", error);
    
    self.downloadInProgress = NO;
    
    // Get the next video
    if (@available(iOS 11.4, *))
    {}
    else
    {
        [self downloadVideoFromQueue];
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        
        if ([self.delegate respondsToSelector:@selector(videoDidFinishDownloadingWithError:)])
        {
            [self.delegate videoDidFinishDownloadingWithError:error];
        }
        
    });
}

- (void)offlineVideoToken:(BCOVOfflineVideoToken)offlineVideoToken
    aggregateDownloadTask:(AVAssetDownloadTask *)downloadtask
            didProgressTo:(NSTimeInterval)progressPercent
        forMediaSelection:(AVMediaSelection *)mediaSelection NS_AVAILABLE_IOS(11_0)
{
    // The specific requested media selected option related to this
    // offline video token has progressed to the specified percent
    NSLog(@"aggregateDownloadTask:didProgressTo:%0.2f for token: %@", progressPercent, offlineVideoToken);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        if ([self.delegate respondsToSelector:@selector(downloadDidProgressTo:)])
        {
            [self.delegate downloadDidProgressTo:progressPercent];
        }
        
    });
}

- (void)offlineVideoToken:(BCOVOfflineVideoToken)offlineVideoToken
didFinishMediaSelectionDownload:(AVMediaSelection *)mediaSelection NS_AVAILABLE_IOS(11_0)
{
    // The specific requested media selected option related to this
    // offline video token has completed downloading
    BCOVOfflineVideoStatus *offlineVideoStatus = [BCOVOfflineVideoManager.sharedManager offlineVideoStatusForToken:offlineVideoToken];
    AVURLAsset *asset = offlineVideoStatus.aggregateDownloadTask.URLAsset;
    NSString *mediaSelectionDescription = [self mediaSelectionDescription:mediaSelection
                                                                 URLAsset:asset];
    NSLog(@"didFinishMediaSelectionDownload:%@ withToken:%@", mediaSelectionDescription, offlineVideoToken);
}

- (void)didDownloadStaticImagesWithOfflineVideoToken:(BCOVOfflineVideoToken)offlineVideoToken
{
    // Called when the thumbnail and poster frame downloads
    // for the specified video token are complete
}

@end

File: Offline/OfflinePlayer/DownloadManager.swift
//
//  DownloadManager.swift
//  OfflinePlayer
//
//  Copyright © 2026 Brightcove, Inc. All rights reserved.
//

import UIKit

import BrightcovePlayerSDK


struct VideoDownload {
    let video: BCOVVideo
    let parameters: [String: Any]
}


final class DownloadManager: NSObject {

    static let shared = DownloadManager()

    class var downloadParameters: [String: Any] {
        // Get base license parameters
        var downloadParameters = DownloadManager.licenseParameters

        guard let window = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .flatMap(\.windows)
                .first(where: \.isKeyWindow),
              let rootViewController = window.rootViewController as? UITabBarController,
              let settingsViewController = rootViewController.settingsViewController else {
            return downloadParameters
        }

        // Add bitrate parameter for the primary download
        let bitrate = settingsViewController.bitrate

        downloadParameters[BCOVOfflineVideoManager.RequestedBitrateKey] = bitrate

        return downloadParameters
    }

    class var licenseParameters: [String: Any] {
        guard let window = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .flatMap(\.windows)
                .first(where: \.isKeyWindow),
              let rootViewController = window.rootViewController as? UITabBarController,
              let settingsViewController = rootViewController.settingsViewController else {
            return .init()
        }

        var licenseParameters: [String: Any] = .init()

        // Generate the license parameters based on the Settings tab
        let isPurchaseLicense = settingsViewController.purchaseLicenseType
        // License details are only needed for FairPlay-protected videos.
        // It's harmless to add it for non-FairPlay videos too.

        if isPurchaseLicense {
            licenseParameters[BCOVFairPlayLicense.PurchaseKey] = true
        } else {
            let rentalDuration = settingsViewController.rentalDuration
            let playDuration = settingsViewController.playDuration

            licenseParameters[BCOVFairPlayLicense.RentalDurationKey] = rentalDuration
            licenseParameters[BCOVFairPlayLicense.PlayDurationKey] = playDuration
        }

        return licenseParameters
    }

    // The download queue.
    // Videos go into the preload queue first.
    // When all preloads are done, videos move to the download queue.
    fileprivate var videoPreloadQueue: [VideoDownload] = .init()
    fileprivate var videoDownloadQueue: [VideoDownload] = .init()

    // Serializes all access to `videoPreloadQueue` and `videoDownloadQueue`,
    // which are read and mutated from both the main queue and a background queue.
    private let queueAccess = DispatchQueue(label: "com.brightcove.offlineplayer.download-queue")

    func doDownload(forVideo video: BCOVVideo) {
        if videoAlreadyProcessing(video) {
            return
        }

        let videoDownload = VideoDownload(video: video,
                                          parameters: DownloadManager.downloadParameters)

        queueAccess.sync {
            videoPreloadQueue.append(videoDownload)
        }

        DispatchQueue.main.async { [self] in
            runPreloadVideoQueue()
        }
    }

    fileprivate func videoAlreadyProcessing(_ video: BCOVVideo) -> Bool {
        // First check to see if the video is in a preload queue
        // videoPreloadQueue is an array of NSDictionary objects,
        // with a BCOVVideo under each "video" key.
        let preloadQueue = queueAccess.sync { videoPreloadQueue }
        for videoDict in preloadQueue {
            if videoDict.video.matches(with: video) {
                UIAlertController.showWith(title: "Video Already in Preload Queue",
                                           message: "The video \(video.localizedName ?? "unknown") is already queued to be preloaded")

                return true
            }
        }

        // First check to see if the video is in a download queue
        // videoDownloadQueue is an array of BCOVVideo objects
        let downloadQueue = queueAccess.sync { videoDownloadQueue }
        for videoDict in downloadQueue {
            if videoDict.video.matches(with: video) {
                UIAlertController.showWith(title: "Video Already in Download Queue",
                                           message: "The video \(video.localizedName ?? "unknown") is already queued to be downloaded")

                return true
            }
        }

        // Next check to see if the video has already been downloaded
        // or is in the process of downloading
        guard let offlineManager = BCOVOfflineVideoManager.sharedManager else {
            return false
        }

        let offlineVideoTokens = offlineManager.offlineVideoTokens

        for offlineVideoToken in offlineVideoTokens {
            guard let offlineVideo = offlineManager.videoObject(fromOfflineVideoToken: offlineVideoToken) else {
                continue
            }

            if offlineVideo.matches(with: video) {
                // If the status is error, alert the user and allow them to retry the download
                if let offlineVideoStatus = offlineManager.offlineVideoStatus(forToken: offlineVideoToken),
                   offlineVideoStatus.downloadState == .error ||
                    offlineVideoStatus.downloadState == .cancelled {
                    UIAlertController.showWith(title: "Video Failed to Download",
                                               message: "The video \(video.localizedName ?? "unknown") previously failed to download or was cancelled, would you like to try again?",
                                               actionTitle: "Retry",
                                               cancelTitle: "Cancel") {

                        offlineManager.deleteOfflineVideo(offlineVideoToken)
                        DownloadManager.shared.doDownload(forVideo: video)
                    }

                    return true
                }

                UIAlertController.showWith(title: "Video Already Downloaded",
                                           message: "The video \(video.localizedName ?? "unknown") is already downloaded (or downloading)")
                return true
            }
        }

        return false
    }

    fileprivate func runPreloadVideoQueue() {
        let videoDownload: VideoDownload? = queueAccess.sync { () -> VideoDownload? in
            guard let videoDownload = videoPreloadQueue.first else {
                return nil
            }

            if let indexOfVideo = videoPreloadQueue.firstIndex(where: { $0.video.matches(with: videoDownload.video) }) {
                videoPreloadQueue.remove(at: indexOfVideo)
            }

            return videoDownload
        }

        guard let videoDownload else {
            DispatchQueue.global(qos: .background).async { [self] in
                downloadVideoFromQueue()
            }

            return
        }

        // Preloading only applies to FairPlay-protected videos.
        // If there's no FairPlay involved, the video is moved on
        // to the video download queue.
        if !videoDownload.video.usesFairPlay {
            queueAccess.sync {
                videoDownloadQueue.append(videoDownload)
            }

            DispatchQueue.main.async { [self] in
                runPreloadVideoQueue()

                NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                                object: videoDownload.video)
            }
        } else {
            guard let offlineManager = BCOVOfflineVideoManager.sharedManager else {
                return
            }

            offlineManager.preloadFairPlayLicense(videoDownload.video,
                                                  parameters: videoDownload.parameters) {
                (offlineVideoToken: String?, error: Error?) in

                DispatchQueue.main.async { [weak self] in
                    guard let self else { return }

                    if let error {
                        // Report any errors
                        UIAlertController.showWith(title: "Video Preload Error\n(\(videoDownload.video.localizedName ?? "unknown"))",
                                                   message: error.localizedDescription)
                    } else {
                        if let offlineVideoToken {
                            print("Preloaded \(offlineVideoToken)")
                        }

                        queueAccess.sync {
                            self.videoDownloadQueue.append(videoDownload)
                        }
                    }

                    runPreloadVideoQueue()

                    NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                                    object: videoDownload.video)
                }
            }
        }
    }

    fileprivate func downloadVideoFromQueue() {
        let videoDownload: VideoDownload? = queueAccess.sync { () -> VideoDownload? in
            guard let videoDownload = videoDownloadQueue.first else {
                return nil
            }

            if let indexOfVideo = videoDownloadQueue.firstIndex(where: { $0.video.matches(with: videoDownload.video) }) {
                videoDownloadQueue.remove(at: indexOfVideo)
            }

            return videoDownload
        }

        guard let videoDownload else {
            return
        }

        guard let offlineManager = BCOVOfflineVideoManager.sharedManager else {
            return
        }

        // Display all available bitrates
        offlineManager.variantBitrates(forVideo: videoDownload.video) {
            (bitrates: [Int]?, error: Error?) in

            print("Variant Bitrates for video: \(videoDownload.video.localizedName ?? "unknown")")

            if let bitrates {
                for bitrate in bitrates {
                    print("\(bitrate)")
                }
            }
        }

        var urlAsset: AVURLAsset?
        do {
            urlAsset = try offlineManager.urlAsset(forVideo: videoDownload.video)
        } catch {}

        if let urlAsset {
            // HLSe (AES-128) streams are not supported for offline playback
            if urlAsset.url.absoluteString.contains("aes128") {
                UIAlertController.showWith(title: "Content Not Supported",
                                           message: "Offline playback is not supported for HLSe content.")
                return
            }
        }

        // If mediaSelections is `nil` the SDK will default to the AVURLAsset's `preferredMediaSelection`
        var mediaSelections = [AVMediaSelection]()

        if let urlAsset {
            mediaSelections = urlAsset.allMediaSelections

            if let legibleMediaSelectionGroup = urlAsset.mediaSelectionGroup(forMediaCharacteristic: .legible),
               let audibleMediaSelectionGroup = urlAsset.mediaSelectionGroup(forMediaCharacteristic: .audible) {

                var counter = 0
                for selection in mediaSelections {
                    let legibleMediaSelectionOption = selection.selectedMediaOption(in: legibleMediaSelectionGroup)
                    let audibleMediaSelectionOption = selection.selectedMediaOption(in: audibleMediaSelectionGroup)

                    let legibleName = legibleMediaSelectionOption?.displayName ?? "nil"
                    let audibleName = audibleMediaSelectionOption?.displayName ?? "nil"

                    print("AVMediaSelection option \(counter) | legible display name: \(legibleName)")
                    print("AVMediaSelection option \(counter) | audible display name: \(audibleName)")
                    counter += 1
                }
            }
        }

        offlineManager.requestVideoDownload(videoDownload.video,
                                            mediaSelections: mediaSelections,
                                            parameters: videoDownload.parameters) {
            (offlineVideoToken: String?, error: Error?) in

            DispatchQueue.main.async {
                if let error {
                    // Report any errors
                    UIAlertController.showWith(title: "Video Download Error",
                                               message: error.localizedDescription)
                }

                NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                                object: videoDownload.video)
            }
        }
    }

    fileprivate class func mediaSelectionDescription(from mediaSelection: AVMediaSelection,
                                                     with urlAsset: AVURLAsset) -> String {
        // Return a string description of the specified Media Selection.
        guard let legibleMSG = urlAsset.mediaSelectionGroup(forMediaCharacteristic: .legible),
              let audibleMSG = urlAsset.mediaSelectionGroup(forMediaCharacteristic: .audible) else {
            return "MediaSelection(n/a)"
        }

        let legibleDisplayName = mediaSelection.selectedMediaOption(in: legibleMSG)?.displayName ?? "-"
        let audibleDisplayName = mediaSelection.selectedMediaOption(in: audibleMSG)?.displayName ?? "-"

        return "MediaSelection(obj:\(mediaSelection), legible:\(legibleDisplayName), audible:\(audibleDisplayName))"
    }
}


// MARK: - BCOVOfflineVideoManagerDelegate

extension DownloadManager: BCOVOfflineVideoManagerDelegate {

    func didCreateSharedBackgroundSesssionConfiguration(_ backgroundSessionConfiguration: URLSessionConfiguration) {
        // Helps prevent downloads from appearing to sometimes stall
        backgroundSessionConfiguration.isDiscretionary = false
    }

    func offlineVideoToken(_ offlineVideoToken: BCOVOfflineVideoToken, aggregateDownloadTask: AVAggregateAssetDownloadTask, didProgressTo progressPercent: TimeInterval, forMediaSelection mediaSelection: AVMediaSelection) {

        // The specific requested media selected option related to this
        // offline video token has progressed to the specified percent
        guard let offlineManager = BCOVOfflineVideoManager.sharedManager,
              let offlineVideo = offlineManager.videoObject(fromOfflineVideoToken: offlineVideoToken) else {

            NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                            object: nil)
            return
        }

        print("aggregateDownloadTask:didProgressTo: \(String(format: "%0.2f", progressPercent)) for token: \(offlineVideoToken)")

        NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                        object: offlineVideo)
    }

    func offlineVideoToken(_ offlineVideoToken: BCOVOfflineVideoToken, didFinishMediaSelectionDownload mediaSelection: AVMediaSelection) {

        // The specific requested media selected option related to this
        // offline video token has completed downloading
        guard let offlineManager = BCOVOfflineVideoManager.sharedManager,
              let offlineVideoStatus = offlineManager.offlineVideoStatus(forToken: offlineVideoToken), let aggregateDownloadTask = offlineVideoStatus.aggregateDownloadTask else {
            return
        }

        let urlAsset = aggregateDownloadTask.urlAsset
        let mediaSelectionDescription = DownloadManager.mediaSelectionDescription(from: mediaSelection,
                                                                                  with: urlAsset)

        print("didFinishMediaSelectionDownload: \(mediaSelectionDescription) for token: \(offlineVideoToken)")
    }

    func offlineVideoToken(_ offlineVideoToken: BCOVOfflineVideoToken, didFinishDownloadWithError error: (any Error)?) {

        // The video has completed downloading
        if let error {
            print("Download finished with error: \(error.localizedDescription)")
        }

        guard let offlineManager = BCOVOfflineVideoManager.sharedManager,
              let offlineVideo = offlineManager.videoObject(fromOfflineVideoToken: offlineVideoToken) else {

            NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                            object: nil)

            return
        }

        NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                        object: offlineVideo)
    }

    func offlineVideoStorageDidChange() {
        NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                        object: nil)
    }
}
